lib/active_model/serializer/reflection.rb

```diff
@@ -17,7 +17,28 @@ class Serializer
     #
     # So you can inspect reflections in your Adapters.
     #
-    Reflection = Struct.new(:name, :options) do
+    Reflection = Struct.new(:name, :options, :block) do
+      delegate :call, to: :reader
+
+      attr_reader :reader
+
+      def initialize(*)
+        super
+        @reader = self.class.build_reader(name, block)
+      end
+
+      def value(instance)
+        call(instance)
+      end
+
+      def self.build_reader(name, block)
+        if block
+          ->(instance) { instance.instance_eval(&block) }
```

Comment:
I'm looking at 0.8 and how JSON API considers associations 'includes' and am thinking that association block should be applied to the association, rather than the serializer.  Otherwise, there's really nothing that ties the block to an association.

So, usage from current impl would be

```diff
has_many :comments do
+  comments.limit(5)
-  limit(5)
end
```

which is also more inline with how the `scope` method works in Rails.  I think this makes the api a lot more clear.  If you want a custom attribute, use the attribute block. If you want a scoped association, I think it the instance_eval should be on the association.

See 

- [0.8](https://github.com/rails-api/active_model_serializers/blob/0-8-stable/lib/active_model/serializer.rb#L115-L142)

```ruby
      def associate(klass, attrs) #:nodoc:
        options = attrs.extract_options!
        self._associations = _associations.dup

        attrs.each do |attr|
          unless method_defined?(attr)
            define_method attr do
              object.send attr
            end
          end

          define_include_method attr

          self._associations[attr] = klass.refine(attr, options)
        end
      end

      def define_include_method(name)
        method = "include_#{name}?".to_sym

        INCLUDE_METHODS[name] = method

        unless method_defined?(method)
          define_method method do
            true
          end
        end
      end
    def include_associations!
      _associations.each_key do |name|
        include!(name) if include?(name)
      end
    end

    def include?(name)
      return false if @options.key?(:only) && !Array(@options[:only]).include?(name)
      return false if @options.key?(:except) && Array(@options[:except]).include?(name)
      send INCLUDE_METHODS[name]
    end

    def include!(name, options={})
```

- [0.9](https://github.com/rails-api/active_model_serializers/blob/0-9-stable/lib/active_model/serializer.rb#L26-L27)

```ruby
      EMBED_IN_ROOT_OPTIONS = [
        :include,
      ]
    def associations(options={})
      associations = self.class._associations
      included_associations = filter(associations.keys)
      associations.each_with_object({}) do |(name, association), hash|
        if included_associations.include? name
        # ...
  end
    def filter(keys)
      if @only
        keys & @only
      elsif @except
        keys - @except
      else
        keys
      end
    end
```